topicscanner.go

```diff
@@ -0,0 +1,115 @@
+package kafka
+
+import (
+	"context"
+	"regexp"
+	"sync"
+	"time"
+)
+
+var (
+	defaultScanningIntervalMS = 10000
+)
+
+type (
+	TopicScannerConfig struct {
+		Brokers            []string
+		TopicRegex         string
+		ScanningIntervalMS int
+	}
+
+	TopicScanner struct {
+		config             TopicScannerConfig
+		currentFoundTopics *[]string
+		closeChan          chan struct{}
+		regex              regexp.Regexp
+		mutex              sync.Mutex
+	}
+)
+
+func NewTopicScanner(config TopicScannerConfig) (*TopicScanner, error) {
+
+	rgx, err := regexp.Compile(config.TopicRegex)
+	if err != nil {
+		return nil, err
+	}
+	//If the scanning interval isn't set properly, scan every 10 seconds
+	if config.ScanningIntervalMS <= 0 {
+		config.ScanningIntervalMS = defaultScanningIntervalMS
+	}
+	return &TopicScanner{
+		currentFoundTopics: new([]string),
+		config:             config,
+		regex:              *rgx,
+		closeChan:          make(chan struct{}),
+	}, nil
+}
+
+func (t *TopicScanner) StartScanning(ctx context.Context) {
+	//Initial fetching of the topics
+	t.updateTopics(ctx)
+	go func() {
+		ticker := time.NewTicker(time.Duration(t.config.ScanningIntervalMS) * time.Millisecond)
+	outer:
+		for {
+			select {
+			case _ = <-ticker.C:
+				t.updateTopics(ctx)
```

Comment:
we probably want "do-while" loop semantics, so we update topics immediately upon startup

that could look something like
```
for {
    t.updateTopics(ctx)

    select {
    case <-ticker.C:
        // noop

    ...

    }
}
```